File: server/tftp_packet.cpp
#include "../server/main.h"
#include "tftp_packet.h"

using namespace std;

TFTP_Packet::TFTP_Packet()	{

	clear();

}

void TFTP_Packet::clear()	{

	current_packet_size = 0;
	
	memset(data, current_packet_size, TFTP_PACKET_MAX_SIZE);

}

unsigned char* TFTP_Packet::getData(int offset) {

	return &(data[offset]);

}

bool TFTP_Packet::copyData(int offset, char* dest, int length) {

	if (offset > this->getSize()) return false;
	if (length < (this->getSize() - offset)) return false; //- out of bounds

	memcpy(dest, &(data[offset]), (this->getSize()-offset));

	return true;

}

void TFTP_Packet::dumpData() {

#ifdef DEBUG

	cout << "\n--------------DATA DUMP---------------------\n";
	cout << "Size: " << current_packet_size << endl;

	for (int i = 0; i < current_packet_size; i++) {
		cout << data[i];
	}

	cout << "\n--------------------------------------------\n\n";

#endif

}

bool TFTP_Packet::addByte(BYTE b) {

	if (current_packet_size >= TFTP_PACKET_MAX_SIZE) {
		return false;
	}

	data[current_packet_size] = (unsigned char)b;
	current_packet_size++;

	return true;

}

bool TFTP_Packet::addWord(WORD w) {
	// pridedam du baitus

	if (!addByte(*(((BYTE*)&w)+1))) return false;

	return (!addByte(*((BYTE*)&w)));

}

bool TFTP_Packet::addString(char* str) {

	int n = strlen(str);

	for (int i=0; i < n; i++) {

		if (!addByte(*(str + i))) { 

			return false;

		}

	}

	return true;

}

bool TFTP_Packet::addMemory(char* buffer, int len) {

	if (current_packet_size + len >= TFTP_PACKET_MAX_SIZE)	{
		cout << "Packet max size exceeded\n";
		return false;
	}

	memcpy(&(data[current_packet_size]), buffer, len);
	current_packet_size += len;

	return true;

}

BYTE TFTP_Packet::getByte(int offset) {

	return (BYTE)data[offset];

}

WORD TFTP_Packet::getWord(int offset) {

	WORD hi = getByte(offset);
	WORD lo = getByte(offset + 1);

	return ((hi<<8)|lo);

}

WORD TFTP_Packet::getNumber() {

	if (this->isData() || this->isACK()) {

		return this->getWord(2);

	} else {

		return 0;

	}

}

bool TFTP_Packet::getString(int offset, char* buffer, int len) {

	if (offset > current_packet_size) return false;
	if (len < current_packet_size - offset) return false;

	memcpy(buffer, &(data[offset]), current_packet_size - offset);

	return true;

}

bool TFTP_Packet::createRRQ(char* filename) {

/*	 2 bytes     string    1 byte     string   1 byte
     Opcode		 Filename  0		  Mode	   0	*/

	clear();
	addWord(TFTP_OPCODE_READ);
	addString(filename);
	addByte(0);
	addString(TFTP_DEFAULT_TRANSFER_MODE);
	addByte(0);

	return true;

}

bool TFTP_Packet::createWRQ(char* filename) {

/*	structure is the same as RRQ  */

	clear();
	addWord(TFTP_OPCODE_WRITE);
	addString(filename);
	addByte(0);
	addString(TFTP_DEFAULT_TRANSFER_MODE);
	addByte(0);

	return true;

}

bool TFTP_Packet::createACK(int packet_num) {

	clear();
	addWord(TFTP_OPCODE_ACK);
	addWord(packet_num);

	return true;

}

bool TFTP_Packet::createData(int block, char* data, int data_size) {

/*	   2 bytes    2 bytes       n bytes
----------------------------------------
DATA  | 03    |   Block #  |    Data    |
---------------------------------------- */

	clear();
	addWord(TFTP_OPCODE_DATA);
	addWord(block);

	addMemory(data, data_size);

	return true;

}

bool TFTP_Packet::createError(int error_code, char* message) {

/*        2 bytes  2 bytes        string    1 byte
          ----------------------------------------
	ERROR | 05    |  ErrorCode |   ErrMsg   |   0  |
          ----------------------------------------  */

	clear();
	addWord(TFTP_OPCODE_ERROR);
	addWord(error_code);
	addString(message);
	addByte(0);

	return true;

}

int TFTP_Packet::getSize() {
	return current_packet_size;
}

bool TFTP_Packet::setSize(int size) {
	if (size <= TFTP_PACKET_MAX_SIZE) {
		current_packet_size = size;
		return true;
	} else {
		return false;
	}
}

bool TFTP_Packet::isRRQ() {

	return (this->getWord(0) == TFTP_OPCODE_READ);

}

bool TFTP_Packet::isWRQ() {

	return (this->getWord(0) == TFTP_OPCODE_WRITE);
}

bool TFTP_Packet::isACK() {

	return (this->getWord(0) == TFTP_OPCODE_ACK);

}


bool TFTP_Packet::isData() {

	return (this->getWord(0) == TFTP_OPCODE_DATA);

}

bool TFTP_Packet::isError() {

	return (this->getWord(0) == TFTP_OPCODE_ERROR);

}

TFTP_Packet::~TFTP_Packet() {
	
	

}

File: java/src/tftp/TFTPPacket.java
package tftp;

import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;

public class TFTPPacket {

	static final char TFTP_OPCODE_READ 	= 1;
	static final char TFTP_OPCODE_WRITE = 2;
	static final char TFTP_OPCODE_DATA 	= 3;
	static final char TFTP_OPCODE_ACK 	= 4;
	static final char TFTP_OPCODE_ERROR = 5;
	
	public static final String TFTP_ERROR_0 = "Not defined, see error message (if any)";
	public static final String TFTP_ERROR_1 = "File not found";
	public static final String TFTP_ERROR_2 = "Access violation";
	public static final String TFTP_ERROR_3 = "Disk full or allocation exceeded";
	public static final String TFTP_ERROR_4 = "Illegal TFTP operation";
	public static final String TFTP_ERROR_5 = "Unknown transfer ID";
	public static final String TFTP_ERROR_6 = "File already exists";
	public static final String TFTP_ERROR_7 = "No such user";
	
	public static final int TFTP_PACKET_MAX_SIZE = 1024;
	public static final int TFTP_PACKET_DATA_SIZE = 512;
	
	static final String TFTP_DEFAULT_TRANSFER_MODE = "octet";
	
	protected int current_packet_size;
	public byte[] data; 
	
	protected int packet_num;
		
	public TFTPPacket() {
		
		current_packet_size = 0;
		packet_num = 0;
		
		data = new byte[TFTP_PACKET_MAX_SIZE];
		
	}
	
	public void clear() {
		
		current_packet_size = 0;
		
		for (int i = 0; i < TFTP_PACKET_MAX_SIZE; i++) data[i] = 0;
		
	}
	
	public int getSize() {
		
		return current_packet_size;
		
	}
	
	public void setSize(int size) throws Exception {
		
		if (size < TFTP_PACKET_MAX_SIZE) {
			current_packet_size = size;
		} else {
			throw new Exception("Packet size exceeded");
		}
		
	}
	
	public void dumpData() {
		
		TFTPUtils.puts("--------------DATA DUMP---------------------");
		TFTPUtils.puts("Packet Size: " + current_packet_size);

		TFTPUtils.puts(new String(data));
		
		TFTPUtils.puts("--------------------------------------------");

	}
	
	public boolean addByte(byte b) {
		
		if (current_packet_size >= TFTP_PACKET_MAX_SIZE) {
			return false;
		}

		data[current_packet_size] = b;
		current_packet_size++;

		return true;
		
	}
	
	public boolean addWord(char w) {

		if (!addByte((byte)((w&0xFF00)>>8))) {
			return false;
		}
		
		return addByte((byte)(w&0x00FF));
		
	}
	
	public boolean addString(String s) {
		
		byte[] b = new byte[s.length()];
		b = s.getBytes();

		for (int i = 0; i < s.length(); i++) {

			if (!addByte(b[i])) { 

				return false;

			}

		}

		return true;
		
	}
	
	public boolean addMemory(byte[] buf, int buf_size) {

		if (current_packet_size + buf_size >= TFTP_PACKET_MAX_SIZE)	{
			TFTPUtils.puts("Packet size exceeded");
			return false;
		}
		
		for (int i = 0; i < buf_size; i++) {
			
			 data[current_packet_size + i] = buf[i];
			
		}
		
		current_packet_size += buf_size;
		
		return false;
		
	}
	
	public byte getByte(int offset) {
		
		return data[offset];
		
	}
	
	public char getWord(int offset) {
		
		return (char)(data[offset] << 8 | data[offset + 1]);
		
	}
	
	public String getString(int offset, int length) throws Exception {

		if (offset > current_packet_size) throw new Exception("getString() out of packet bounds");

		if (length < current_packet_size - offset) throw new Exception("getString() length is out of packet bounds");

		String output = new String();

		for (int i = offset; i < offset + length; i++) {
			
			if (data[i] == 0) break; //zero-terminated

			output += (char)data[i];

		}
		
		return output;
		
	}
	
	public char getPacketNumber() {
		
		return (isData() || isACK()) ? getWord(2) : 0;
		
	}
	
	public byte[] getData(int offset) {
		
		if (!isData()) return null;
		
		byte[] data_part = new byte[getSize() - 4];
		
		for (int i = offset; i < current_packet_size; i++) {
			data_part[i-offset] = data[i];
		}
		
		return data_part;
		
	}
	
	
	
	public void createRRQ(String filename) {
		
		clear();
		addWord(TFTP_OPCODE_READ);
		addString(filename);
		addByte((byte)0);
		addString(TFTP_DEFAULT_TRANSFER_MODE);
		addByte((byte)0);

	}
	
	public void createWRQ(String filename) {
		
		clear();
		addWord(TFTP_OPCODE_WRITE);
		addString(filename);
		addByte((byte)0);
		addString(TFTP_DEFAULT_TRANSFER_MODE);
		addByte((byte)0);

	}
	
	public void createACK(char packet_num) {
		
		clear();
		addWord(TFTP_OPCODE_ACK);
		addWord(packet_num);

	}	

	public void createData(int block, byte[] data, int data_size) {
		
		clear();
		addWord(TFTP_OPCODE_DATA);
		addWord((char)block);
		addMemory(data, data_size);

	}

	public void createError(int error_code, String message) {
		
		clear();
		addWord(TFTP_OPCODE_ERROR);
		addWord((char)error_code);
		addString(message);
		addByte((byte)0);

	}
	
	public boolean isRRQ() {
		
		return ((int)getWord(0) == TFTP_OPCODE_READ);		
		
	}
	
	public boolean isWRQ() {

		return ((int)getWord(0) == TFTP_OPCODE_WRITE);
		
	}
	
	public boolean isACK() {
		
		return ((int)getWord(0) == TFTP_OPCODE_ACK);
		
	}
	
	public boolean isData() {
		
		return ((int)getWord(0) == TFTP_OPCODE_DATA);
		
	}
	
	public boolean isError() {
		
		return ((int)getWord(0) == TFTP_OPCODE_ERROR);
		
	}
	
	public boolean sendPacket(BufferedOutputStream out) {
		
		try {
			out.write(data, 0, getSize());
			out.flush();
			return true;
		} catch (Exception e) {
			TFTPUtils.puts("Exception in sendPacket()");
			return false;
		}
		
	}
	
	public boolean getPacket(BufferedInputStream in) {
		
		clear();

   		int bytes_read = 0;
   		
   		try {
   			bytes_read = in.read(data, 0, TFTP_PACKET_MAX_SIZE);
   			
   			if (bytes_read == -1) {
   	   			return false;
   	   		}
   			
   			setSize(bytes_read);
   			
   		} catch (Exception e) {
   			TFTPUtils.puts("Exception in getPacket()");
   			return false;
   		}
   		
   		return true;
		
	}
	

}
